### ISSUE TITLE
Error when saving config for some specific strings

### ISSUE THREAD

**Issue by CyrilleB79** [nvaccess/nvda#17839]
### Steps to reproduce:
* Open advanced settings
* In Regular expression for text paragraph navigation, add `'''"""` at the end of the regexp.
* Press OK
* Press `NVDA+control+C` to save the config.

### Actual behavior:
When the user tries to manually save the config (`NVDA+control+C` or NVDA menu -> Save configuration), the configuration is not actually saved. An error is logged, but no error message is displayed in case saving the config fails. Thus the user is not informed that the config cannot be saved anymore, unless they are running a test version or they activate the error sounds in advanced settings.

It's because `configobj` cannot store any type of string.

### Expected behavior:
When the user tries to manually save the config (`NVDA+control+C` or NVDA menu -> Save configuration), at least an error message error is displayed in case saving the config fails.

Note that this string is not expected in the regexp for normal usage, so a more precise information is not needed.

### NVDA logs, crash dumps and other attachments:
```ERROR - unhandled exception (10:20:20.050) - MainThread (12992):
Traceback (most recent call last):
  File "wx\core.pyc", line 3427, in <lambda>
  File "gui\blockAction.pyc", line 79, in funcWrapper
  File "gui\__init__.pyc", line 220, in onSaveConfigurationCommand
  File "config\__init__.pyc", line 746, in save
  File "config\__init__.pyc", line 736, in _writeProfileToFile
  File "configobj\__init__.pyc", line 2040, in write
  File "configobj\__init__.pyc", line 2042, in write
  File "configobj\__init__.pyc", line 1951, in _write_line
  File "configobj\__init__.pyc", line 1796, in _quote
  File "configobj\__init__.pyc", line 1035, in _get_triple_quote
configobj.ConfigObjError: Value cannot be safely quoted: '((?<=\\w)(?<!\\d)["”»)]?[.…]{1,3}["”»)]?(\\[\\d+\\])*(?=[\\r\\n \xa0]|$)|[?!]|[．！？：；])"""\'\'\''
```
#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
2024.4.2
#### Windows version:
Windows 10 22H2 (AMD64) build 19045.5487
#### Name and version of other software in use when reproducing the issue:
N/A
#### Other information about your system:
N/A
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
N/A

**Comment by CyrilleB79** [comment]
@seanbudd, I have opened this issue upon request of @SaschaCowley (https://github.com/nvaccess/nvda/pull/17827#issuecomment-2731215734).

Though, it is resolved by #17827.

Or are you expecting something else?

**Comment by seanbudd** [comment]
Oh I see it has been closed, thanks

**Comment by SaschaCowley** [comment]
Arguably this hasn't been fixed, we just report the issue to the user. We really should be attempting to escape the quotation marks in these cases, if possible

**Comment by CyrilleB79** [comment]
@SaschaCowley wrote:
> Arguably this hasn't been fixed, we just report the issue to the user. We really should be attempting to escape the quotation marks in these cases, if possible

We could probably fix this issue by encoding/decoding the string with json if you absolutely want to be able to save it.
Though, it would means a config upgrade for this regexp.

Also, this issue was already present in NVDA 2024.4. But iIt's worth noting that many more new string options are being added during 2025.1 dev cycle: NVDA remote host, NVDA remote key, NVDA update mirror, Add-on store update mirror, and maybe other ones... And thus they should probably be upgraded too.

But strings that are invalid for config obj are also invalid or at least unexpected for URLs, expected regexp, host name, and for key name. Rather than trying to save them absolutely, couldn't we rather just forbid invalid strings in these fields and issue an error message in case the user tries to use it? This would have no real impact on users since they were not able to save them in any case before that.
And there is no point in developing a specific save mechanism for a situation that never happens in real life. We may specifically think to saving strings with json the day when we have an edit field that can contain such a patologic string in for a real use case, e.g. if we have an edit field able to contain Python code.

@SaschaCowley or @seanbudd what are your thoughts?

**Comment by SaschaCowley** [comment]
@CyrilleB79 I can already see a use case for having single and double quotes in the paragraph regex: supporting single-quoted strings. 

I also don't know that this would require a config upgrade, if we made escaping these characters global. I'm not sure how we would achieve that on a technical level though.

**Comment by CyrilleB79** [comment]
> [@CyrilleB79](https://github.com/CyrilleB79) I can already see a use case for having single and double quotes in the paragraph regex: supporting single-quoted strings.

Today, you can already add single quote in this regexp, i.e. have both single and double quote in this paragraph regexp.
The problem is if you want to add both tripled single quote and tripled double quote in this regexp. That's why I have indicated that it does not correspond to any known use case in today's GUI.

> I also don't know that this would require a config upgrade, if we made escaping these characters global. I'm not sure how we would achieve that on a technical level though.

What do you mean with making escaping these characters global?
If some characters are escaped in the future and are not today, I guess the only way to do this is to add a sublayer when saving/reading the config i.e. change the way configobj perform these actions. I have difficulty to figure how this could be achieved without a config upgrade, i.e. keeping the old configs compatibility.

**Comment by CyrilleB79** [comment]
> > What do you mean with making escaping these characters global? If some characters are escaped in the future and are not today, I guess the only way to do this is to add a sublayer when saving/reading the config i.e. change the way configobj perform these actions. I have difficulty to figure how this could be achieved without a config upgrade, i.e. keeping the old configs compatibility.
> 
> Hmm, I was imagining some sort of system where we escape the minimum number of characters required to make the string compatible. For example, rather than writing `"""`, we write `""\"` or `""\x22`, which would not require changing currently legal strings. I fear this would be impractical though.

Do you mean adding a layer when saving/reading the config file which would act as follows:
1. If the string can already be saved today, the future new layer keeps it unchanged
2. IF the string cannot be saved today, the new layer would add some sort of escaping so that it can be saved.

In this case, we should ensure that the strings produced in the config file in case 2 could never have been produced with today's NVDA. It does not seem possible to me. Even if it was, it seems to be a very specific implementation for a case that never happens for now.
If one day we have to save Python code in the config, I'd rather encode this specificstring in json (or base64 if readability is not required) and do not define a global layered save mechanism.

Though, to close this PR, we can just add a security so that the error message is produced before you save to `config.conf`, i.e. when you press OK in the settings dialog, not when you try to save the config to disk, i.e. when you press `NVDA+control+c` or when NVDA saves on exit.

**Comment by SaschaCowley** [comment]
@CyrilleB79 that is what I was imagining, yes. I agree with you that it is impractical and would probably be quite brittle. I think the current state of affairs is fine.
